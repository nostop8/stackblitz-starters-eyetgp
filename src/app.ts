import "express-async-errors";
import express, { Request, Response, NextFunction } from "express";
import json from "body-parser";
import {
  addProductToCart,
  authUser,
  getProductsSortedAlphabetically,
} from "./services/dummy.service";
import { CartPayload, RequestWithUser } from "./types";
import { UnauthorizedHttpError } from "./errors/unauthorized-error";
import { errorMiddleware } from "./middlewares/error";
import { authorizeMiddleware } from "./middlewares/authorize";
import { BadRequestError } from "./errors/bad-request-error";

const app = express();

app.use(json());

app.get("/", (req: Request, res: Response) => {
  res.send("Hello, World!");
});

app.get("/products", async (req: Request, res: Response) => {
  const products = await getProductsSortedAlphabetically();
  res.send(products);
});

app.post("/products", async (req: Request, res: Response) => {
  // no among the readme tasks, skipping
  res.send();
});

app.post("/login", async (req: Request, res: Response) => {
  const { username, password } = req.body;
  const user = await authUser({ username, password });
  res.send(user);
});

app.post(
  "/cart",
  authorizeMiddleware,
  async (req: RequestWithUser, res: Response) => {
    const { userId } = req;

    if (!userId) {
      throw new UnauthorizedHttpError("User unauthorized");
    }

    const { productId } = req.body as unknown as CartPayload;

    if (!productId) {
      throw new BadRequestError("Product id is required");
    }

    addProductToCart({ userId, productId });
    res.send();
  }
);

app.use((req: Request, res: Response, next: NextFunction) => {
  res.status(404).send();
});

app.use(errorMiddleware);

export default app;
