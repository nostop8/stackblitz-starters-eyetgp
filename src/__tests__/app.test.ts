import app from "../app";
import supertest from "supertest";
import products from "./__stubs__/products.json";
import user from "./__stubs__/user.json";
import {
  createUserCart,
  fetchProducts,
  fetchUserCarts,
  postAuthLogin,
  updateUserCart,
} from "../services/dummy.api";
import {
  HTTP_STATUS_BAD_REQUEST,
  HTTP_STATUS_OK,
  HTTP_STATUS_UNAUTHORIZED,
} from "../consts";

jest.mock("../services/dummy.api");

const request = supertest(app);

describe("GET /", () => {
  it('should return "Hello, World!"', async () => {
    const response = await request.get("/");
    expect(response.status).toBe(HTTP_STATUS_OK);
    expect(response.text).toBe("Hello, World!");
  });
});

describe("GET /products", () => {
  it("should return products sorted by A-Z", async () => {
    (fetchProducts as jest.Mock).mockResolvedValue({ data: products });

    const response = await request.get("/products").expect(HTTP_STATUS_OK);

    expect(response.body).toMatchSnapshot();
  });

  it("should return error if anything wrong", async () => {
    (fetchProducts as jest.Mock).mockRejectedValue(new Error("error"));

    await request.get("/products").expect(HTTP_STATUS_BAD_REQUEST);
  });
});

describe("POST /login", () => {
  it("should return user", async () => {
    (postAuthLogin as jest.Mock).mockResolvedValue({ data: user });

    const response = await request
      .post("/login")
      .send({
        username: "john",
        password: "password",
      })
      .expect(HTTP_STATUS_OK);

    expect(response.body).toMatchSnapshot();
  });

  it("should return unauthorized error", async () => {
    (postAuthLogin as jest.Mock).mockResolvedValue({
      status: HTTP_STATUS_BAD_REQUEST,
    });

    const response = await request
      .post("/login")
      .send({
        username: "john",
        password: "password",
      })
      .expect(HTTP_STATUS_UNAUTHORIZED);

    expect(response.body).toMatchSnapshot();
  });
});

describe("POST /cart", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should return error if user unauthorized", async () => {
    await request.post("/cart").expect(HTTP_STATUS_UNAUTHORIZED);
  });

  it("should return error if token missing user id", async () => {
    await request
      .post("/cart")
      .set({ Authorization: "Bearer token" })
      .expect(HTTP_STATUS_UNAUTHORIZED);
  });

  it('should return error if "productId" missing', async () => {
    await request
      .post("/cart")
      .set({ Authorization: `Bearer ${user.token}` })
      .expect(HTTP_STATUS_BAD_REQUEST);
  });

  it("shoud add product to new user cart and return OK", async () => {
    (fetchUserCarts as jest.Mock).mockResolvedValue({ data: { carts: [] } });

    await request
      .post("/cart")
      .set({ Authorization: `Bearer ${user.token}` })
      .send({ productId: 2 })
      .expect(HTTP_STATUS_OK);

    expect(fetchUserCarts).toBeCalledWith({ userId: user.id });
    expect(createUserCart).toBeCalledWith({
      userId: user.id,
      products: [{ id: 2, quantity: 1 }],
    });
    expect(updateUserCart).not.toBeCalled();
  });

  it("should add product to first available user cart and return OK", async () => {
    (fetchUserCarts as jest.Mock).mockResolvedValue({
      data: { carts: [{ id: 1 }] },
    });

    await request
      .post("/cart")
      .set({ Authorization: `Bearer ${user.token}` })
      .send({ productId: 3 })
      .expect(HTTP_STATUS_OK);

    expect(fetchUserCarts).toBeCalledWith({ userId: user.id });
    expect(updateUserCart).toBeCalledWith({
      id: 1,
      userId: user.id,
      products: [{ id: 3, quantity: 1 }],
    });
    expect(createUserCart).not.toBeCalled();
  });
});
