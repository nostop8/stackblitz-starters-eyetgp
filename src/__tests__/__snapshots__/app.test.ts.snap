// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /products should return products sorted by A-Z 1`] = `
[
  {
    "description": "HP Pavilion 15-DK1056WM Gaming Laptop 10th Gen Core i5, 8GB, 256GB SSD, GTX 1650 4GB, Windows 10",
    "id": 10,
    "price": 1099,
    "thumbnail": "https://i.dummyjson.com/data/products/10/thumbnail.jpeg",
    "title": "HP Pavilion 15-DK1056WM",
  },
  {
    "description": "Huawei’s re-badged P30 Pro New Edition was officially unveiled yesterday in Germany and now the device has made its way to the UK.",
    "id": 5,
    "price": 499,
    "thumbnail": "https://i.dummyjson.com/data/products/5/thumbnail.jpg",
    "title": "Huawei P30",
  },
  {
    "description": "Infinix Inbook X1 Ci3 10th 8GB 256GB 14 Win10 Grey – 1 Year Warranty",
    "id": 9,
    "price": 1099,
    "thumbnail": "https://i.dummyjson.com/data/products/9/thumbnail.jpg",
    "title": "Infinix INBOOK",
  },
  {
    "description": "An apple mobile which is nothing like apple",
    "id": 1,
    "price": 549,
    "thumbnail": "https://i.dummyjson.com/data/products/1/thumbnail.jpg",
    "title": "iPhone 9",
  },
  {
    "description": "SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip with ...",
    "id": 2,
    "price": 899,
    "thumbnail": "https://i.dummyjson.com/data/products/2/thumbnail.jpg",
    "title": "iPhone X",
  },
  {
    "description": "MacBook Pro 2021 with mini-LED display may launch between September, November",
    "id": 6,
    "price": 1749,
    "thumbnail": "https://i.dummyjson.com/data/products/6/thumbnail.png",
    "title": "MacBook Pro",
  },
  {
    "description": "Style and speed. Stand out on HD video calls backed by Studio Mics. Capture ideas on the vibrant touchscreen.",
    "id": 8,
    "price": 1499,
    "thumbnail": "https://i.dummyjson.com/data/products/8/thumbnail.jpg",
    "title": "Microsoft Surface Laptop 4",
  },
  {
    "description": "OPPO F19 is officially announced on April 2021.",
    "id": 4,
    "price": 280,
    "thumbnail": "https://i.dummyjson.com/data/products/4/thumbnail.jpg",
    "title": "OPPOF19",
  },
  {
    "description": "Samsung Galaxy Book S (2020) Laptop With Intel Lakefield Chip, 8GB of RAM Launched",
    "id": 7,
    "price": 1499,
    "thumbnail": "https://i.dummyjson.com/data/products/7/thumbnail.jpg",
    "title": "Samsung Galaxy Book",
  },
  {
    "description": "Samsung's new variant which goes beyond Galaxy to the Universe",
    "id": 3,
    "price": 1249,
    "thumbnail": "https://i.dummyjson.com/data/products/3/thumbnail.jpg",
    "title": "Samsung Universe 9",
  },
]
`;

exports[`POST /login should return unauthorized error 1`] = `
{
  "message": "Invalid credentials",
}
`;

exports[`POST /login should return user 1`] = `
{
  "avatar": "https://robohash.org/autquiaut.png",
  "firstName": "Jeanne",
  "lastName": "Halvorson",
  "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTUsInVzZXJuYW1lIjoia21pbmNoZWxsZSIsImVtYWlsIjoia21pbmNoZWxsZUBxcS5jb20iLCJmaXJzdE5hbWUiOiJKZWFubmUiLCJsYXN0TmFtZSI6IkhhbHZvcnNvbiIsImdlbmRlciI6ImZlbWFsZSIsImltYWdlIjoiaHR0cHM6Ly9yb2JvaGFzaC5vcmcvYXV0cXVpYXV0LnBuZyIsImlhdCI6MTY5NTU4NjczNCwiZXhwIjoxNjk1NTkwMzM0fQ.GcaZGjjXm6wgenGhh-13vGEY211q_Jkc-V3XB6E1XwE",
  "username": "kminchelle",
}
`;
